src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -1222,6 +1222,64 @@ void f_constraint(
                                  inverse_spatial_metric, gauge_function, phi,
                                  inverse_spacetime_metric);
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+Scalar<DataType> constraint_energy_normalization(
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_psi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_pi,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& d_phi,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const double psi_multiplier, const double pi_multipiler,
+    const double phi_multiplier) noexcept {
+  auto normalization = make_with_value<Scalar<DataType>>(d_pi, 0.0);
+  constraint_energy_normalization<SpatialDim, Frame, DataType>(
+      &normalization, d_psi, d_pi, d_phi, inverse_spatial_metric,
+      psi_multiplier, pi_multipiler, phi_multiplier);
+  return normalization;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void constraint_energy_normalization(
+    gsl::not_null<Scalar<DataType>*> normalization,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_psi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_pi,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& d_phi,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const double psi_multiplier, const double pi_multiplier,
+    const double phi_multiplier) noexcept {
+  if (get_size(get(*normalization)) !=
+      get_size(get_size(get<0, 0, 0>(d_psi)))) {
+    *normalization = Scalar<DataType>{get_size(get<0, 0, 0>(d_psi))};
+  }
+  get(*normalization) = psi_multiplier * inverse_spatial_metric.get(0, 0) *
+                            d_psi.get(0, 0, 0) * d_psi.get(0, 0, 0) +
+                        pi_multiplier * inverse_spatial_metric.get(0, 0) *
+                            d_pi.get(0, 0, 0) * d_pi.get(0, 0, 0);
+
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t j = 0; j < SpatialDim; ++j) {
+      for (size_t a = 0; a < SpatialDim + 1; ++a) {
+        for (size_t c = 0; c < SpatialDim + 1; ++c) {
+          if (not(i == 0 and j == 0 and a == 0 and c == 0)) {
```

Comment:
`UNLIKELY(not...)`

Reply:
wait, but it's the other way that's unlikely. If the code is correct as written, the if should pass on all but one of the loops. So, to do the unlikely, the if should be put around the next block with two more loops and the not removed.

Reply:
yeah do LIKELY